plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,153 @@
+import io.airlift.json.ObjectMapperProvider;
+import io.airlift.log.Logger;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoInputStream;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorSplitSource;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_FILESYSTEM_ERROR;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_S3_CROSS_REGION_UNSUPPORTED;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftUnloadSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    private final CompletableFuture<Void> resultSetFuture;
+    private final Connection connection;
+    private final Statement statement;
+    private final String unloadOutputPath;
+    private final TrinoFileSystem fileSystem;
+
+    private Optional<Location> manifestLocation;
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(ExecutorService executor, Connection connection, PreparedStatement statement, String queryFragmentId, String unloadOutputPath, TrinoFileSystem fileSystem)
+    {
+        requireNonNull(executor, "executor is null");
+        this.connection = requireNonNull(connection, "connection is null");
+        this.statement = requireNonNull(statement, "statement is null");
+        this.unloadOutputPath = requireNonNull(unloadOutputPath, "unloadOutputPath is null");
+        verify(unloadOutputPath.endsWith("/"), "unloadOutputPath must end with '/'.");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        resultSetFuture = CompletableFuture.runAsync(() -> {
+            log.debug("Executing: %s", statement);
+            try {
+                // Exclusively set readOnly to false to avoid query failing with "ERROR: transaction is read-only".
+                connection.setReadOnly(false);
+                long beginTs = System.currentTimeMillis();
+                statement.execute(); // Return value of `statement.execute()` is not useful as it always return false.
+                log.info("UNLOAD command for %s query took %sms", queryFragmentId, System.currentTimeMillis() - beginTs);
+            }
+            catch (SQLException e) {
```

Comment:
I think we will leak connection on exceptions.
The connection should be managed in one place rather than moving across split manager and split source.
You can just do everything with the connection here in one try/finally block which closes the connection rather than some things in split manager and some here.